### ISSUE TITLE
The Sims 4 mods are not loaded

### ISSUE THREAD

**Issue by Miresoft** [ModOrganizer2/modorganizer#1952]
## The problem:
Mods are enabled in MO2 and are visible in the VFS, but do not appear in the Mods window when launching the game.

## To Reproduce:

1. Install and enable a mod for The Sims 4
2. Launch the game using the Run button in MO2
3. Open the Mods window in the options.

## Environment:

* Mod Organizer Version that exhibits the issue: 2.5.0
* Last Mod Organizer Version that did not exhibit the issue (if applicable):
* Desktop OS/version used to run Mod Organizer: Windows 10, in french

## Details:
At first, the game was installed in french and I had to edit the GameDataPath in game_sims4.py to "%DOCUMENTS%/Electronic Arts/Les Sims 4/Mods" for MO2 to open without error. Mods were not working so I reinstalled the game in english, but the mods are still not visible when loading de game.

### USVFS:
https://gist.github.com/Miresoft/492363fa9e8b2033a85ebf41254b7db2

### MO Interface:
https://gist.github.com/Miresoft/95cfc54d2a57ed2b3cd0fdd7a22b84f5

**Comment by vertigomaster** [comment]
Been experiencing this same issue as well; it seems MO2 is having difficulty following the process chain from TS4 to the EADesktop app, so it thinks the game has closed. Issue still persists if EADesktop.exe is closed, as EADesktop is seemingly launched by some weird EA background service. Uncertain how EA games detecting when it is launched by the launcher.

Curious if this same issue is occuring with other games on EA's platform

**Comment by DJLegends1011** [comment]
same issue as well

**Comment by qwertydude567** [comment]
I'm also having the same issue, as well as mental breakdown as a result.

**Comment by keplerk** [comment]
+1 I have the same issue, the virtual env doesn't seems to get injected into the game :(

**Comment by qwertydude567** [comment]
Geesh... OK, so yay, I got it to work.  I will show you how to also get it to work (hopefully).  I say geesh, because it was a pain in the ass and I'm pissed off that it was this much trouble and that I had to waste a lot of time to get it to work.  It's the god-damned EA App - the stupid app does not have advanced launch options, which means it doesn't allow the user (you) to set custom launch commands when loading software through it.  But its superior precurser, Origin App, does.  When you run the game, it forces one of EA's apps to run - either the Origin App or the EA App.  Normally running the Origin App tries to force an upgrade to the EA App, but you can hack it so that it doesn't and you can keep using the old (and better) Origin App (with the advanced launch options that you need).   I also have Steam installed, which also tries to force an upgrade of the EA App, but I can prevent it from doing so easily...  Anyway.  Here's what I did:

Go here: https://github.com/p0358/Fuck_off_EA_App and follow their instructions on downloading and installing the Origin App.  Afterwards, you should be able to load up the Origin App launcher without being hassled to install the EA App.  Then you go and load up Mod Organizer and go to your Sims 4 "instance" (if you installed Mod Organizer that way), otherwise maybe it's just set automatically to The Sims 4?  Then click on the "Shortcut" dropdown just below the "Run" button.  Select "Desktop".  Find the shortcut on your desktop and right-click it.  Choose the "Copy as path" option.  Now go into the Origin App.  Go into "My Game Library".  Right-click The Sims 4 thumbnail.  Choose "Game Properties"  Select "ADVANCED LAUNCH OPTIONS".  Click on the line, hold CTRL+V on your keyboard to paste in the shortcut path, then after the end quote, type a space and then "%command&" (without the quotes).  Save it.  Then go back into Mod Organizer and click "Run".  For me, Steam tries to install the EA App at this point, but I just close the popup about it (while I present my middle finger to my screen) and the game loads up with the Mod Organizer virtual mod setup like we all expected it to from the start of this fiasco before we saw that it wasn't working with the shitty EA App crap.

Hopefully the Mod Organizer programmers can take my info and create a better solution that Mod Organizer can automatically do for us, but if you want it to work for now, hopefully these instructions work for you.

**Comment by qwertydude567** [comment]
Actually, lol, you don't even need to do the advanced launch command at all.  I just got rid of it.  Just clicking Run in Mod Organizer works on it's own.  I was wrong about the advanced launch option requirement.  I don't know what it is about the EA App, but it just fucks everything up.  The Origin App doesn't.  All you need to do is just obliterate the EA App and use the Origin App instead.  That's it.

**Comment by keplerk** [comment]
What if i have the game on Steam ? it means it has to go through 2 different apps ?

**Comment by Holt59** [comment]
> What if i have the game on Steam ? it means it has to go through 2 different apps ?

If you have the game on Steam, you're already going through 2 apps because Steam simply launches the EA Desktop App, which is the issue here because the EA App then starts the game in a separate unrelated process and the code that should be injected by MO2 isn't. The solution above is to revert back to Origin that does not spawn an unrelated process and thus keeps proper injection (you're still going through Origin to launch the game, but it works).

**Comment by keplerk** [comment]
> If you have the game on Steam, you're already going through 2 apps because Steam simply launches the EA Desktop App, which is the issue here because the EA App then starts the game in a separate unrelated process and the code that should be injected by MO2 isn't. The solution above is to revert back to Origin that does not spawn an unrelated process and thus keeps proper injection (you're still going through Origin to launch the game, but it works).

In resume, follow the steps you described, and everything should works?
Btw is normal that my antivirus detects the installer as a possible threat?

**Comment by Holt59** [comment]
> Btw is normal that my antivirus detects the installer as a possible threat?

I'm not the one that proposed the solution, but yes, should work.

If you're talking about the installer from https://github.com/p0358/Fuck_off_EA_App, I would not be surprised if it's flagged as a virus. This does not come from MO2, so I cannot say if it's legit or not.

**Comment by keplerk** [comment]
> If you're talking about the installer from https://github.com/p0358/Fuck_off_EA_App, I would not be surprised if it's flagged as a virus. This does not come from MO2, so I cannot say if it's legit or not.

Yeah, i checked out the source code and the false positive is due the app does some patches but seems okish.
Thanks for the help !

**Comment by qwertydude567** [comment]
No problem.  The method works, for now.  EA may go through efforts to try to prevent people from hacking their push to the EA App.  It would be nice if in the meantime if someone with the MO development team could develop some kind of work-around still using the EA App.

I think the false positive in antivirus (I got it too) is because hardly any people who allow antivirus software to report back data have run the program - so it has a very low usage rating and thus antivirus can't verify the safety of it.  I honestly didn't know if it was 100% safe, but it did work... lol

Follow the steps I described without the details I gave regarding the advanced launch options, because they aren't required.  I decided to keep them in there just so you can see what I also tried to do, and then know that I determined that part wasn't necessary for me to get it to work.

**Comment by keplerk** [comment]
> No problem. The method works, for now. EA may go through efforts to try to prevent people from hacking their push to the EA App. It would be nice if in the meantime if someone with the MO development team could develop some kind of work-around still using the EA App.
> 
> I think the false positive in antivirus (I got it too) is because hardly any people who allow antivirus software to report back data have used the file(s) - so with a very low usage rating the antivirus can't verify the safety of it. I honestly didn't know if it was 100% safe, but it did work... lol
> 
> Follow the steps I described without the details I gave regarding the advanced launch options, because they aren't required. I decided to keep them in there just so you can see what I also tried to do, and then know that I determined that part wasn't necessary for me to get it to work.

Indeed, i just did the ini file modification plus the installation and check the option on the same. Everything worked like a charm.

**Comment by Silarn** [comment]
This may be the 'best' solution for now. EA is trying to kill Origin. Unsure how long that workaround is going to be viable.

**Comment by keplerk** [comment]
> This may be the 'best' solution for now. EA is trying to kill Origin. Unsure how long that workaround is going to be viable.

Hope long enough for MO2 team to come out with a solution for EA app :(

**Comment by mnwachukwu** [comment]
Just wanted to drop by to say that using the Origin app might work for some, but for me, who has never logged into Origin before, the offline mode wont work because there is no logged in profile for it to check against, essentially leaving me dead in the water.

I am hoping that a fix can be found soon!

**Comment by qwertydude567** [comment]
Here's a workaround.  Find a reliable source for a cracked game executable that will allow you to run the game offline and completely skip loading the EA App or Origin App.

**Comment by Rolothe5th** [comment]
looks like EA just added offline support!

**Comment by Silarn** [comment]
Is there a launch arg you can use to trigger that?

**Comment by xieve** [comment]
> Just wanted to drop by to say that using the Origin app might work for some, but for me, who has never logged into Origin before today, the offline mode wont work because there is no logged in profile for it to check against, essentially leaving me dead in the water. At least, I surmise that's what happening when I can't log in via offline mode but can via the EA App.
> 
> I am hoping that a fix can be found soon!

I just installed the Origin app and https://github.com/p0358/Fuck_off_EA_App and logged in without a problem. Didn't even have to uninstall the EA App, closing it was enough. Now when I start TS4 via MO2, it launches through Origin and everything works fine.

**Comment by JRCx2Gaming** [comment]
I figured out how to fix it.

First go to (\Documents\Electronic Arts\The Sims 4) and delete the "mods" folder.
Open CMD (command prompt) as Admin.
Type this command, but with your MO2 folder location

mklink /D "C:\Users\*username*\OneDrive\Documents\Electronic Arts\The Sims 4\Mods" "D:\Modding\Sims 4 - MO2\mods"

This creates an Symbolic link between the two folders to where you can use MO2 to mod Sims 4.

**Comment by JRCx2Gaming** [comment]
The link works, but there's 2 issues I've found. 
1. If you want to disable a mod you have to completely remove it.
2. When installing a mod, in the Data menu on the right plane it doubles everything. (Which is just a visual thing, doesn't actually double everything)

**Comment by vertigomaster** [comment]
I'm wondering if that's an incidental bug or a semi-intentional security tactic on EA's to prevent these kinds of hooks, perhaps for anti-cheat purposes?

In any case, that's certainly an interesting find. 

Do you know what the virtual address that caused the crash was? Do we think it was more of a windows permissions issue or something controlled by the EA app?

**Comment by TsunaMoo** [comment]
Your suspicions seem correct as EA recently has shown that they do not wish any of their software to run on linux or through any kind of virtualisation including VMs.
I'd have to make another crash dump but I recall the address accessed was a strange one like 000000dx

**Comment by docmop** [comment]
Depending on the chosen language, the games mods folder does **not** match `%DOCUMENTS%/Electronic Arts/The Sims 4/Mods`. If the game is installed in German for example it's `%DOCUMENTS%/Electronic Arts/Die Sims 4/Mods`.

You can not derive the path automatically from the install directory of the game either as that is still `[...]/The Sims 4` irrespective of chosen language.

One solution to get the data panel to work, is manually adjusting `GameDataPath` in `game_sims4.py`.

Could we expose the GameDataPath to the settings frontend?
